titus-server-gateway/src/main/java/com/netflix/titus/gateway/endpoint/v3/grpc/TitusGatewayGrpcServer.java

```diff
@@ -112,67 +115,69 @@ public int getPort() {
-                throw new RuntimeException(e);
-            }
-            LOG.info("Started gRPC server on port {}.", port);
+        if (started.getAndSet(true)) {
+            return;
+        }
+        this.port = config.getPort();
+        ServerBuilder<?> serverBuilder = configure(ServerBuilder.forPort(port).executor(grpcCallbackExecutor))
+                .addService(ServerInterceptors.intercept(
+                        healthService,
+                        createInterceptors(HealthGrpc.getServiceDescriptor())
+                ))
+                .addService(ServerInterceptors.intercept(
+                        jobManagementService,
+                        createInterceptors(JobManagementServiceGrpc.getServiceDescriptor())
+                ))
+                .addService(ServerInterceptors.intercept(
+                        evictionService,
+                        createInterceptors(EvictionServiceGrpc.getServiceDescriptor())
+                ))
+                .addService(ServerInterceptors.intercept(
+                        agentManagementService,
+                        createInterceptors(AgentManagementServiceGrpc.getServiceDescriptor())
+                ))
+                .addService(ServerInterceptors.intercept(
+                        appAutoScalingService,
+                        createInterceptors(AutoScalingServiceGrpc.getServiceDescriptor())
+                ))
+                .addService(ServerInterceptors.intercept(
+                        schedulerService,
+                        createInterceptors(SchedulerServiceGrpc.getServiceDescriptor())
+                )).addService(
+                        ServerInterceptors.intercept(
+                                reactorServerFactory.apply(
+                                        MachineServiceGrpc.getServiceDescriptor(),
+                                        reactorMachineGrpcService
+                                ),
+                                createInterceptors(MachineServiceGrpc.getServiceDescriptor())
+                        )
+                )
+                .addService(ProtoReflectionService.newInstance());
+
+        if (config.getLoadBalancerGrpcEnabled()) {
+            serverBuilder.addService(ServerInterceptors.intercept(
+                    loadBalancerService,
+                    createInterceptors(LoadBalancerServiceGrpc.getServiceDescriptor())));
         }
+        this.server = serverBuilder.build();
+
+        LOG.info("Starting gRPC server on port {}.", port);
+        try {
+            this.server.start();
+            this.port = server.getPort();
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
+        LOG.info("Started gRPC server on port {}.", port);
     }
 
     @PreDestroy
     public void shutdown() {
-        if (!server.isShutdown()) {
-            long timeoutMs = config.getShutdownTimeoutMs();
+        if (server.isShutdown()) {
```

Comment:
Can this code be extracted into a single place instead of it being duplicated in each server?

Reply:
it is already extract in `AbstractTitusGrpcServer`, but these three older server classes (Master, Gateway, Federation) are not using it yet.

Reply:
(and I didn't want to make that change in this PR)